<a href="https://colab.research.google.com/github/egipot/ZTMPy/blob/main/256_HackerNewsProject2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# import requests
# from bs4 import BeautifulSoup
# import pprint

# res = requests.get('https://news.ycombinator.com/news')
# res2 = requests.get('https://news.ycombinator.com/news?p=2')
# soup = BeautifulSoup(res.text, 'html.parser')
# soup2 = BeautifulSoup(res2.text, 'html.parser')

# links = soup.select('.titleline > a') #heads up! .storylink changed to .titleline
# subtext = soup.select('.subtext')
# links2 = soup2.select('.titleline > a') #heads up! .storylink changed to .titleline
# subtext2 = soup2.select('.subtext')

# mega_links = links + links2
# mega_subtext = subtext + subtext2

# def sort_stories_by_votes(hnlist):
#   return sorted(hnlist, key= lambda k:k['votes'], reverse=True)

# def create_custom_hn(links, subtext):
#   hn = []
#   for idx, item in enumerate(links):
#     title = item.getText()
#     href = item.get('href', None)
#     vote = subtext[idx].select('.score')
#     if len(vote):
#       points = int(vote[0].getText().replace(' points', ''))
#       if points > 99:
#         hn.append({'title': title, 'link': href, 'votes': points})
#   return sort_stories_by_votes(hn)

# pprint.pprint(create_custom_hn(mega_links, mega_subtext))


In [1]:
import requests
from bs4 import BeautifulSoup
import pprint

In [2]:
res = requests.get('https://news.ycombinator.com/news')
res2 = requests.get('https://news.ycombinator.com/news?p=2')
soup = BeautifulSoup(res.text, 'html.parser')
soup2 = BeautifulSoup(res2.text, 'html.parser')

In [3]:

links = soup.select('.titleline > a') #heads up! .storylink changed to .titleline
subtext = soup.select('.subtext')
links2 = soup2.select('.titleline > a') #heads up! .storylink changed to .titleline
subtext2 = soup2.select('.subtext')

In [10]:
links

[<a href="https://www.hillelwayne.com/post/graph-types/">The hunt for the missing data type</a>,
 <a href="https://www.anthropic.com/news/claude-3-family">Claude 3 model family</a>,
 <a href="https://astro.sshh.io/">Show HN: Astro App</a>,
 <a href="https://www.ubicloud.com/blog/improving-network-performance-with-linux-flowtables">Improving Network Performance with Linux Flowtables</a>,
 <a href="https://www.surfertoday.com/surfing/the-woman-who-coined-the-expression-surfing-the-internet">The women who coined the expression 'surfing the internet' (2019)</a>,
 <a href="https://opus-codec.org/demo/opus-1.5/">Opus 1.5 released: Opus gets a machine learning upgrade</a>,
 <a href="item?id=39591972">Launch HN: Just words (YC W24) – Optimize your product's copy for user growth</a>,
 <a href="https://www.nytimes.com/2024/03/04/science/moon-landing-sideways-gravity.html" rel="nofollow">Why it's so challenging to land upright on the moon</a>,
 <a href="https://mainlymiles.com/2024/03/03/singapor

In [11]:
links2

[<a href="https://www.youtube.com/watch?v=UyKtxsdI0z8" rel="nofollow">Hermeus Hypersonic Aircraft – Real Engineering [video]</a>,
 <a href="https://ieeexplore.ieee.org/abstract/document/6319640">A windowed graph Fourier transform (2012)</a>,
 <a href="https://www.osar.fr/notes/motionblur/">Motion blur all the way down (2022)</a>,
 <a href="https://deprogrammaticaipsum.com/twenty-years-is-nothing/">Twenty years is nothing</a>,
 <a href="https://voussoir.net/writing/css_for_printing">CSS for printing to paper</a>,
 <a href="https://www-cdn.anthropic.com/de8ba9b01c9ab7cbabf5c33b80b7bbc618857627/Model_Card_Claude_3.pdf" rel="nofollow">The Claude 3 Model Family: Opus, Sonnet, Haiku [pdf]</a>,
 <a href="https://www.nature.com/articles/d41586-024-00616-5">More than 2M research papers have disappeared from the Internet</a>,
 <a href="https://www.gresham.ac.uk/sites/default/files/2021-02-09-1300-HART_FictionalWorlds-T.pdf">Mathematical journeys into fictional worlds (2021) [pdf]</a>,
 <a href="

In [12]:
subtext

[<td class="subtext"><span class="subline">
 <span class="score" id="score_39592444">189 points</span> by <a class="hnuser" href="user?id=todsacerdoti">todsacerdoti</a> <span class="age" title="2024-03-04T16:42:14"><a href="item?id=39592444">2 hours ago</a></span> <span id="unv_39592444"></span> | <a href="hide?id=39592444&amp;goto=news">hide</a> | <a href="item?id=39592444">86 comments</a> </span>
 </td>,
 <td class="subtext"><span class="subline">
 <span class="score" id="score_39590666">605 points</span> by <a class="hnuser" href="user?id=marc__1">marc__1</a> <span class="age" title="2024-03-04T14:08:51"><a href="item?id=39590666">5 hours ago</a></span> <span id="unv_39590666"></span> | <a href="hide?id=39590666&amp;goto=news">hide</a> | <a href="item?id=39590666">348 comments</a> </span>
 </td>,
 <td class="subtext"><span class="subline">
 <span class="score" id="score_39591529">149 points</span> by <a class="hnuser" href="user?id=sshh12">sshh12</a> <span class="age" title="2024-03

In [13]:
subtext2

[<td class="subtext"><span class="subline">
 <span class="score" id="score_39594490">3 points</span> by <a class="hnuser" href="user?id=tuanx5">tuanx5</a> <span class="age" title="2024-03-04T19:02:40"><a href="item?id=39594490">17 minutes ago</a></span> <span id="unv_39594490"></span> | <a href="hide?id=39594490&amp;goto=news%3Fp%3D2">hide</a> | <a href="item?id=39594490">1 comment</a> </span>
 </td>,
 <td class="subtext"><span class="subline">
 <span class="score" id="score_39588977">42 points</span> by <a class="hnuser" href="user?id=teleforce">teleforce</a> <span class="age" title="2024-03-04T10:49:36"><a href="item?id=39588977">8 hours ago</a></span> <span id="unv_39588977"></span> | <a href="hide?id=39588977&amp;goto=news%3Fp%3D2">hide</a> | <a href="item?id=39588977">3 comments</a> </span>
 </td>,
 <td class="subtext"><span class="subline">
 <span class="score" id="score_39586828">294 points</span> by <a class="hnuser" href="user?id=azeemba">azeemba</a> <span class="age" title="2

In [7]:
def create_custom_hn(links, subtext):
  hn = []
  for idx, item in enumerate(links):
    title = item.getText()
    href = item.get('href', None)
    vote = subtext[idx].select('.score')
    if len(vote):
      points = int(vote[0].getText().replace(' points', ''))
      if points > 99:
        hn.append({'title': title, 'link': href, 'votes': points})
  return hn

print(create_custom_hn(links, subtext))

[{'title': 'The hunt for the missing data type', 'link': 'https://www.hillelwayne.com/post/graph-types/', 'votes': 189}, {'title': 'Claude 3 model family', 'link': 'https://www.anthropic.com/news/claude-3-family', 'votes': 605}, {'title': 'Show HN: Astro App', 'link': 'https://astro.sshh.io/', 'votes': 149}, {'title': 'Singapore Airlines Concorde', 'link': 'https://mainlymiles.com/2024/03/03/singapore-airlines-concorde-full-story/', 'votes': 144}, {'title': 'I made a bad camera lens from some old glasses [video]', 'link': 'https://www.youtube.com/watch?v=-C6AcRnUmbE', 'votes': 119}, {'title': 'Rust for Embedded Systems: Current state, challenges and open problems', 'link': 'https://arxiv.org/abs/2311.05063', 'votes': 106}, {'title': 'French court issues damages award for violation of GPL', 'link': 'https://heathermeeker.com/2024/02/17/french-court-issues-damages-award-for-violation-of-gpl/', 'votes': 797}, {'title': 'New 13- and 15‑inch MacBook Air with M3 chip', 'link': 'https://www.a

In [9]:
def create_custom_hn(links, subtext):
  hn = []
  for idx, item in enumerate(links):
    title = item.getText()
    href = item.get('href', None)
    vote = subtext[idx].select('.score')
    hn.append({'title': title, 'link': href, 'votes': vote})
  return hn

print(create_custom_hn(links, subtext))

[{'title': 'The hunt for the missing data type', 'link': 'https://www.hillelwayne.com/post/graph-types/', 'votes': [<span class="score" id="score_39592444">189 points</span>]}, {'title': 'Claude 3 model family', 'link': 'https://www.anthropic.com/news/claude-3-family', 'votes': [<span class="score" id="score_39590666">605 points</span>]}, {'title': 'Show HN: Astro App', 'link': 'https://astro.sshh.io/', 'votes': [<span class="score" id="score_39591529">149 points</span>]}, {'title': 'Improving Network Performance with Linux Flowtables', 'link': 'https://www.ubicloud.com/blog/improving-network-performance-with-linux-flowtables', 'votes': [<span class="score" id="score_39592771">51 points</span>]}, {'title': "The women who coined the expression 'surfing the internet' (2019)", 'link': 'https://www.surfertoday.com/surfing/the-woman-who-coined-the-expression-surfing-the-internet', 'votes': [<span class="score" id="score_39592993">49 points</span>]}, {'title': 'Opus 1.5 released: Opus gets a

In [18]:
def create_custom_hn(links, subtext):
  hn = []
  for idx, item in enumerate(links):
    title = item.getText()
    href = item.get('href', None)
    vote = subtext[idx].select('#score_39592444')
    hn.append({'title': title, 'link': href, 'votes': vote})
  return hn

print(create_custom_hn(links, subtext))

[{'title': 'The hunt for the missing data type', 'link': 'https://www.hillelwayne.com/post/graph-types/', 'votes': [<span class="score" id="score_39592444">189 points</span>]}, {'title': 'Claude 3 model family', 'link': 'https://www.anthropic.com/news/claude-3-family', 'votes': []}, {'title': 'Show HN: Astro App', 'link': 'https://astro.sshh.io/', 'votes': []}, {'title': 'Improving Network Performance with Linux Flowtables', 'link': 'https://www.ubicloud.com/blog/improving-network-performance-with-linux-flowtables', 'votes': []}, {'title': "The women who coined the expression 'surfing the internet' (2019)", 'link': 'https://www.surfertoday.com/surfing/the-woman-who-coined-the-expression-surfing-the-internet', 'votes': []}, {'title': 'Opus 1.5 released: Opus gets a machine learning upgrade', 'link': 'https://opus-codec.org/demo/opus-1.5/', 'votes': []}, {'title': "Launch HN: Just words (YC W24) – Optimize your product's copy for user growth", 'link': 'item?id=39591972', 'votes': []}, {'

[{'title': 'The hunt for the missing data type', 'link': 'https://www.hillelwayne.com/post/graph-types/', 'votes': [<span class="score" id="score_39592444">189 points</span>]}, {'title': 'Claude 3 model family', 'link': 'https://www.anthropic.com/news/claude-3-family', 'votes': []}, {'title': 'Show HN: Astro App', 'link': 'https://astro.sshh.io/', 'votes': []}, {'title': 'Improving Network Performance with Linux Flowtables', 'link': 'https://www.ubicloud.com/blog/improving-network-performance-with-linux-flowtables', 'votes': []}, {'title': "The women who coined the expression 'surfing the internet' (2019)", 'link': 'https://www.surfertoday.com/surfing/the-woman-who-coined-the-expression-surfing-the-internet', 'votes': []}, {'title': 'Opus 1.5 released: Opus gets a machine learning upgrade', 'link': 'https://opus-codec.org/demo/opus-1.5/', 'votes': []}, {'title': "Launch HN: Just words (YC W24) – Optimize your product's copy for user growth", 'link': 'item?id=39591972', 'votes': []}, {'title': "Why it's so challenging to land upright on the moon", 'link': 'https://www.nytimes.com/2024/03/04/science/moon-landing-sideways-gravity.html', 'votes': []}, {'title': 'Singapore Airlines Concorde', 'link': 'https://mainlymiles.com/2024/03/03/singapore-airlines-concorde-full-story/', 'votes': []}, {'title': 'Intel Previews Sierra Forest with 288 E-Cores, Announces Granite Rapids-D', 'link': 'https://www.anandtech.com/show/21276/intel-previews-sierra-forest-with-288-e-cores-announces-granite-rapids-d-for-2025-launch-at-mwc-2024', 'votes': []}, {'title': 'Skio (YC S20) – Subscriptions for Shopify, ReCharge migrations –\xa0Is Hiring', 'link': 'https://skio.com/careers/', 'votes': []}, {'title': 'I made a bad camera lens from some old glasses [video]', 'link': 'https://www.youtube.com/watch?v=-C6AcRnUmbE', 'votes': []}, {'title': 'Rust for Embedded Systems: Current state, challenges and open problems', 'link': 'https://arxiv.org/abs/2311.05063', 'votes': []}, {'title': "U.S. Air Force Evaluating Pivotal's Blackfly Ultralight EVTOL Aircraft", 'link': 'https://www.futureflight.aero/news-article/2024-02-27/us-air-force-evaluating-pivotals-blackfly-ultralight-evtol-aircraft', 'votes': []}, {'title': 'French court issues damages award for violation of GPL', 'link': 'https://heathermeeker.com/2024/02/17/french-court-issues-damages-award-for-violation-of-gpl/', 'votes': []}, {'title': "MacPad: I created the hybrid Mac-iPad laptop and tablet that Apple won't make", 'link': 'https://www.macstories.net/stories/macpad-how-i-created-the-hybrid-mac-ipad-laptop-and-tablet-that-apple-wont-make/', 'votes': []}, {'title': 'Borrow checking without lifetimes', 'link': 'https://smallcultfollowing.com/babysteps/blog/2024/03/04/borrow-checking-without-lifetimes/', 'votes': []}, {'title': 'New 13- and 15‑inch MacBook Air with M3 chip', 'link': 'https://www.apple.com/newsroom/2024/03/apple-unveils-the-new-13-and-15-inch-macbook-air-with-the-powerful-m3-chip/', 'votes': []}, {'title': 'What Is a Schur Decomposition? (2022)', 'link': 'https://nhigham.com/2022/05/11/what-is-a-schur-decomposition/', 'votes': []}, {'title': 'Revy – proof-of-concept time-travel debugger for the Bevy game engine', 'link': 'https://github.com/rerun-io/revy', 'votes': []}, {'title': 'Yuzu emulator developers settle Nintendo lawsuit, pay $2.4M in damages', 'link': 'https://twitter.com/oatmealdome/status/1764704580724576465', 'votes': []}, {'title': 'Nvidia bans using translation layers for CUDA software to run on other chips', 'link': 'https://www.tomshardware.com/pc-components/gpus/nvidia-bans-using-translation-layers-for-cuda-software-to-run-on-other-chips-new-restriction-apparently-targets-zluda-and-some-chinese-gpu-makers', 'votes': []}, {'title': 'What the CEO wants you to know (2023)', 'link': 'https://commoncog.com/what-the-ceo-wants-you-to-know/', 'votes': []}, {'title': 'Velato: A programming language where source code must be a valid MIDI music file', 'link': 'http://www.velato.net/', 'votes': []}, {'title': 'What we learned from C++ atomics and memory model standardization [video]', 'link': 'https://www.youtube.com/watch?v=Ss7gIs2-hzk', 'votes': []}, {'title': 'Show HN: Workflow orchestrator in Golang', 'link': 'https://github.com/harshadmanglani/polaris', 'votes': []}, {'title': 'A broken vintage piano turned into an analog hybrid of 20 instruments', 'link': 'https://www.themusicman.uk/broken-vintage-piano/', 'votes': []}, {'title': 'A mathematician who finds poetry in math and math in poetry', 'link': 'https://www.quantamagazine.org/the-theorist-who-sees-math-in-art-music-and-writing-20240112/', 'votes': []}, {'title': 'A computational approach to emergence in complex multi-level systems', 'link': 'https://arxiv.org/abs/2402.09090', 'votes': []}, {'title': 'Comparing Postgres Managed Services: AWS, Azure, GCP and Supabase', 'link': 'https://blog.peerdb.io/comparing-postgres-managed-services-aws-azure-gcp-and-supabase', 'votes': []}]


In [21]:
def create_custom_hn(links, subtext):
  hn = []
  for idx, item in enumerate(links):
    title = item.getText()
    href = item.get('href', None)
    vote = subtext[idx].select('.score')
    if len(vote):
      points = int(vote[0].getText().replace(' points', ''))
      print(points)
    hn.append({'title': title, 'link': href, 'votes': vote})
  return hn

print(create_custom_hn(links, subtext))

189
605
149
51
49
98
30
7
144
42
119
106
8
797
82
16
185
72
30
106
192
19
60
26
5
53
75
79
38
[{'title': 'The hunt for the missing data type', 'link': 'https://www.hillelwayne.com/post/graph-types/', 'votes': [<span class="score" id="score_39592444">189 points</span>]}, {'title': 'Claude 3 model family', 'link': 'https://www.anthropic.com/news/claude-3-family', 'votes': [<span class="score" id="score_39590666">605 points</span>]}, {'title': 'Show HN: Astro App', 'link': 'https://astro.sshh.io/', 'votes': [<span class="score" id="score_39591529">149 points</span>]}, {'title': 'Improving Network Performance with Linux Flowtables', 'link': 'https://www.ubicloud.com/blog/improving-network-performance-with-linux-flowtables', 'votes': [<span class="score" id="score_39592771">51 points</span>]}, {'title': "The women who coined the expression 'surfing the internet' (2019)", 'link': 'https://www.surfertoday.com/surfing/the-woman-who-coined-the-expression-surfing-the-internet', 'votes': [<span c

In [23]:
def create_custom_hn(links, subtext):
  hn = []
  for idx, item in enumerate(links):
    title = item.getText()
    href = item.get('href', None)
    vote = subtext[idx].select('.score')
    if len(vote):
      points = int(vote[0].getText().replace(' points', ''))
      # print(points)
    hn.append({'title': title, 'link': href, 'votes': points})
  return hn

print(create_custom_hn(links, subtext))

[{'title': 'The hunt for the missing data type', 'link': 'https://www.hillelwayne.com/post/graph-types/', 'votes': 189}, {'title': 'Claude 3 model family', 'link': 'https://www.anthropic.com/news/claude-3-family', 'votes': 605}, {'title': 'Show HN: Astro App', 'link': 'https://astro.sshh.io/', 'votes': 149}, {'title': 'Improving Network Performance with Linux Flowtables', 'link': 'https://www.ubicloud.com/blog/improving-network-performance-with-linux-flowtables', 'votes': 51}, {'title': "The women who coined the expression 'surfing the internet' (2019)", 'link': 'https://www.surfertoday.com/surfing/the-woman-who-coined-the-expression-surfing-the-internet', 'votes': 49}, {'title': 'Opus 1.5 released: Opus gets a machine learning upgrade', 'link': 'https://opus-codec.org/demo/opus-1.5/', 'votes': 98}, {'title': "Launch HN: Just words (YC W24) – Optimize your product's copy for user growth", 'link': 'item?id=39591972', 'votes': 30}, {'title': "Why it's so challenging to land upright on th

In [26]:
import pprint


In [27]:

def create_custom_hn(links, subtext):
  hn = []
  for idx, item in enumerate(links):
    title = item.getText()
    href = item.get('href', None)
    vote = subtext[idx].select('.score')
    if len(vote):
      points = int(vote[0].getText().replace(' points', ''))
      # print(points)
    hn.append({'title': title, 'link': href, 'votes': points})
  return hn

pprint.pprint(create_custom_hn(links, subtext))

[{'link': 'https://www.hillelwayne.com/post/graph-types/',
  'title': 'The hunt for the missing data type',
  'votes': 189},
 {'link': 'https://www.anthropic.com/news/claude-3-family',
  'title': 'Claude 3 model family',
  'votes': 605},
 {'link': 'https://astro.sshh.io/',
  'title': 'Show HN: Astro App',
  'votes': 149},
 {'link': 'https://www.ubicloud.com/blog/improving-network-performance-with-linux-flowtables',
  'title': 'Improving Network Performance with Linux Flowtables',
  'votes': 51},
 {'link': 'https://www.surfertoday.com/surfing/the-woman-who-coined-the-expression-surfing-the-internet',
  'title': "The women who coined the expression 'surfing the internet' (2019)",
  'votes': 49},
 {'link': 'https://opus-codec.org/demo/opus-1.5/',
  'title': 'Opus 1.5 released: Opus gets a machine learning upgrade',
  'votes': 98},
 {'link': 'item?id=39591972',
  'title': "Launch HN: Just words (YC W24) – Optimize your product's copy for "
           'user growth',
  'votes': 30},
 {'link'

In [29]:
#add a filter to links with 100 (or more) votes

def create_custom_hn(links, subtext):
  hn = []
  for idx, item in enumerate(links):
    title = item.getText()
    href = item.get('href', None)
    vote = subtext[idx].select('.score')
    if len(vote):
      points = int(vote[0].getText().replace(' points', ''))
      # print(points)
      if points > 99:
        hn.append({'title': title, 'link': href, 'votes': points})
  return hn

pprint.pprint(create_custom_hn(links, subtext))

[{'link': 'https://www.hillelwayne.com/post/graph-types/',
  'title': 'The hunt for the missing data type',
  'votes': 189},
 {'link': 'https://www.anthropic.com/news/claude-3-family',
  'title': 'Claude 3 model family',
  'votes': 605},
 {'link': 'https://astro.sshh.io/',
  'title': 'Show HN: Astro App',
  'votes': 149},
 {'link': 'https://mainlymiles.com/2024/03/03/singapore-airlines-concorde-full-story/',
  'title': 'Singapore Airlines Concorde',
  'votes': 144},
 {'link': 'https://www.youtube.com/watch?v=-C6AcRnUmbE',
  'title': 'I made a bad camera lens from some old glasses [video]',
  'votes': 119},
 {'link': 'https://arxiv.org/abs/2311.05063',
  'title': 'Rust for Embedded Systems: Current state, challenges and open '
           'problems',
  'votes': 106},
 {'link': 'https://heathermeeker.com/2024/02/17/french-court-issues-damages-award-for-violation-of-gpl/',
  'title': 'French court issues damages award for violation of GPL',
  'votes': 797},
 {'link': 'https://www.apple.com